### PR TITLE
[63117] Neural API Support

### PR DESCRIPTION
**Pull request by mrashed-dev**
# Description
This PR brings support for the Nylas [Neural API](https://developer.nylas.com/docs/intelligence/). The Neural API exposes endpoints that perform intelligent analysis on items in your Nylas account. This change will enable users to access the features of the Nylas Neural API and provides some utility functions to help improve the user's workflow.

# Usage
To use [Sentiment Analysis](https://developer.nylas.com/docs/intelligence/sentiment-analysis/):
```
// To perform sentiment analysis on a message, pass in the message ID:
const messageAnalysis = await nylas.neural.sentimentAnalysisMessage(MESSAGE_ID);

// To perform sentiment analysis on just text, pass in a string:
const textAnalysis = await nylas.neural.sentimentAnalysisText("Hi, thank you so much for reaching out! We can catch up tomorrow.");
```

To use [Signature Extraction](https://developer.nylas.com/docs/intelligence/signature-extraction/):
```
const signature = await nylas.neural.extractSignature(MESSAGE_ID);

// The method also accepts two optional parameters
// parseContact, a boolean for whether Nylas should parse the contact from the signature (API defaults to true)
// options, an object of options that can be enabled for the Neural endpoint, of type NeuralMessageOptions:
const options = {
    ignore_links: false,
    ignore_images: false,
    ignore_tables: false,
    remove_conclusion_phrases: false,
    images_as_markdowns: false
}

const signature = await nylas.neural.extractSignature(MESSAGE_ID, true, options);
```
and to parse the contact and convert it to the standard Nylas contact object:
```
const contact = signature.contacts.toContactObject();
```

To use [Clean Conversations](https://developer.nylas.com/docs/intelligence/clean-conversations/):
```
const convo = await nylas.neural.cleanConversation(MESSAGE_ID);

// You can also pass in an object of options that can be enabled for the Neural endpoint, of type NeuralMessageOptions
const convo = await nylas.neural.cleanConversation(MESSAGE_ID, options);
```
and to extract images from the result:
```
await convo.extractImages();
```

To use [Optical Character Recognition](https://developer.nylas.com/docs/intelligence/optical-charecter-recognition/):
```
const ocr = await nylas.neural.ocrRequest(FILE_ID);

// This endpoint also supports a second, optional parameter for an array specifying the pages that the user wants analyzed:
const ocr = await nylas.neural.ocrRequest(FILE_ID, [2,3]);
```

To use [Categorizer](https://developer.nylas.com/docs/intelligence/categorizer/)
```
let cat = await nylas.neural.categorize(MESSAGE_ID);

// You can also send a request to recategorize the message:
cat = await cat.reCategorize("conversation");
```

# License
I confirm that this contribution is made under the terms of the MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Review by philrenaud — APPROVED** [review]

File: src/models/neural-clean-conversation.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import File from './file';
+
+export default class NeuralCleanConversation extends Message {
+  conversation?: string;
+  modelVersion?: string;
+
+  extractImages(): Promise<File[]> {
+    const f: File[] = [];
+    if (this.conversation) {
+      const regex = /[(']cid:(.)*[)']/g;
```
Comment (sdwvit) [inline src/models/neural-clean-conversation.ts:12]: would be nice to move this on the top of the file

**Review by tjperry07 — APPROVED** [review]
Love the examples in the comments! Makes it so much easier for me to write documentation.

File: src/models/neural-clean-conversation.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      const fileIds = this.conversation.match(regex);
+      if (fileIds) {
+        fileIds.forEach(async id => {
+          const parsedId = id.substring(5, id.length - 1);
```
Comment (sdwvit) [inline src/models/neural-clean-conversation.ts:16]: would be nice to have a comment about how id looks like

File: src/models/neural-sentiment-analysis.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import RestfulModel from './restful-model';
+
+export default class NeuralSentimentAnalysis extends RestfulModel {
+  account_id?: string;
+  sentiment?: string;
+  sentimentScore?: number;
+  processedLength?: number;
+  text?: string;
+
+  fromJSON(json: object): this {
```
Comment (sdwvit) [inline src/models/neural-sentiment-analysis.ts:11]: how about 
```suggestion
  fromJSON(json: unknown): this {
```

Comment (sdwvit) [reply src/models/neural-sentiment-analysis.ts:11]: or typeof this

File: src/models/neural.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        ...body,
+        ...options,
+      };
+    }
+    if (parseContact) {
+      body['parse_contact'] = parseContact;
+    }
+
+    return this._request(path, body).then(json => {
+      if (Array.isArray(json)) {
```
Comment (sdwvit) [inline src/models/neural.ts:49]: in which cases there may be more elements?

**Review by sdwvit — APPROVED** [review]

**Review by arjunkalburgi — APPROVED** [review]

File: src/models/neural-sentiment-analysis.ts (hunk 3 above)
Comment (mrashed-dev) [reply src/models/neural-sentiment-analysis.ts:11]: @sdwvit just curious, why going from typing object to unknown?

File: src/models/neural.ts (hunk 4 above)
Comment (mrashed-dev) [reply src/models/neural.ts:49]: According to the docs, never, but sometimes the API response that gets returned is in the form of an Array of just 1  element. Here's an example: https://developer.nylas.com/docs/intelligence/signature-extraction/#parse-signature-response

File: src/models/neural-clean-conversation.ts (hunk 1 above)
Comment (mrashed-dev) [reply src/models/neural-clean-conversation.ts:12]: Good catch!

File: src/models/neural-clean-conversation.ts (hunk 2 above)
Comment (mrashed-dev) [reply src/models/neural-clean-conversation.ts:17]: Good idea.... I should probably have better comments in general

File: src/models/neural-sentiment-analysis.ts (hunk 3 above)
Comment (sdwvit) [reply src/models/neural-sentiment-analysis.ts:11]: object is as broad as any in this case. Unknown just means you don't know the type. 
Not the best idea to have either of those. 

How about: 
```suggestion
  fromJSON(json: Partial<typeof this>): this {
```

Comment (sdwvit) [reply src/models/neural-sentiment-analysis.ts:11]: or anything like that, that imrpoves type

Comment (mrashed-dev) [reply src/models/neural-sentiment-analysis.ts:11]: I love the suggestion! Actually it turns out I don't even need this piece of code, it's redundant all it does is call the super method and the class `extends RestfulModel` so there's no need for it.

**Comment by mrashed-dev** [comment]
Thanks for everyone's reviews! I will get a new release with this functionality out sometime tomorrow! 🎉

**Pull request merged**